### PR TITLE
PR: Optimise keyPressEvent in the editor

### LINKED ISSUE #11412: Spyder 4.0.1 editing *really* laggy
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [x] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [x] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
After updating from the beta4 to 4.0.1 I am experiencing very laggy keyboard responses in the editor. I have turned off Kite and this doesn't seem to solve the problem, so it appears to be part of the Spyder default (rather than Kite addon) configuration

### What steps reproduce the problem?

1. Update spyder to latest 4.0.1
2. Open a new session
3. Start editing a ~1500 line *.py script and watch as key presses are not synced with characters appearing in the editor

### What is the expected output? What do you see instead?

Instant keyboard, character appearance

## Versions

* Spyder version: 4.0.1
* Python version: 3.7.6
* Qt version: 5.9.7
* PyQt version: 5.9.2
* Operating System name/version: MacOS 10.14.6

### Dependencies
```
applaunchservices >=0.1.7    :  0.2.1 (OK)
atomicwrites >=1.2.0         :  1.3.0 (OK)
chardet >=2.0.0              :  3.0.4 (OK)
cloudpickle >=0.5.0          :  1.2.2 (OK)
diff_match_patch >=20181111  :  20181111 (OK)
intervaltree                 :  None (OK)
IPython >=4.0                :  7.11.1 (OK)
jedi =0.14.1                 :  0.14.1 (OK)
nbconvert >=4.0              :  5.6.1 (OK)
numpydoc >=0.6.0             :  0.9.2 (OK)
pexpect >=4.4.0              :  4.7.0 (OK)
pickleshare >=0.4            :  0.7.5 (OK)
psutil >=0.3                 :  5.6.7 (OK)
pygments >=2.0               :  2.5.2 (OK)
pylint >=0.25                :  2.4.4 (OK)
pyls >=0.31.2;<0.32.0        :  0.31.2 (OK)
zmq >=17                     :  18.1.0 (OK)
qdarkstyle >=2.7             :  2.7 (OK)
qtawesome >=0.5.7            :  0.6.0 (OK)
qtconsole >=4.6.0            :  4.6.0 (OK)
qtpy >=1.5.0                 :  1.9.0 (OK)
rtree >=0.8.3                :  0.8.3 (OK)
sphinx >=0.6.6               :  2.3.1 (OK)
spyder_kernels >=1.8.1;<2.0.0:  1.8.1 (OK)
watchdog                     :  None (OK)
cython >=0.21                :  None (NOK)
matplotlib >=2.0.0           :  None (NOK)
numpy >=1.7                  :  1.17.5 (OK)
pandas >=0.13.1              :  None (NOK)
scipy >=0.17.0               :  None (NOK)
sympy >=0.7.3                :  None (NOK)
```

### LINKED ISSUE #11495: Flags are misaligned when scrolling if code is folded
The scrolling behaviour is broken on master:
broken_master

_Originally posted by @impact27 in https://github.com/spyder-ide/spyder/pull/11488#issuecomment-583304164_

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

`keyPressEvent` are slow on large file. one of the reasons is that for each key press, the entire file is read to search for cells to highlight. Here these cells are cached.

Also, the flags in the scroll bar are computed for the entire document, and not just for the visible lines. This is corrected as well.

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #11495
Fixes #11412

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by ccordoba12** [comment]
> Also, the flags in the scroll bar are computed for the entire document, and not just for the visible lines. This is corrected as well.

I don't understand this part because the scrollbar reflects the contents of the entire file.

**Comment by impact27** [comment]
> > Also, the flags in the scroll bar are computed for the entire document, and not just for the visible lines. This is corrected as well.
> 
> I don't understand this part because the scrollbar reflects the contents of the entire file.

Only when the scroll slider is visible, when it is not the flags reflect only the current lines.

**Comment by ccordoba12** [comment]
> Only when the scroll slider is visible, when it is not the flags reflect only the current lines.

Please post a screenshot because I still don't understand very well what you're trying to achieve.

**Comment by impact27** [comment]
moving
fixed

**Comment by impact27** [comment]
The scrolling behaviour is broken on master:
broken_master

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 1]
```diff
@@ -2214,17 +2217,6 @@ def redo(self):
             self.sig_text_was_inserted.emit()
             self.skip_rstrip = False
 
-    def get_block_data(self, block):
```
Comment (impact27) [inline spyder/plugins/editor/widgets/codeeditor.py]: Unused functions

**Comment by goanpeca** [comment]
@impact27 but thwat happens if the user hits enter? the location of all flags on the scroll bar needs to be updated, is this the case with your changes?

**Comment by impact27** [comment]
> @impact27 but thwat happens if the user hits enter? the location of all flags on the scroll bar needs to be updated, is this the case with your changes?

Yes

File: spyder/plugins/editor/extensions/snippets.py [hunk 2]
```diff
@@ -700,17 +700,9 @@ def cursor_changed(self, line, col):
             self.inserting_snippet = False
             return
 
-        ignore_calls = {'undo', 'redo'}
         node, nearest_snippet, _ = self._find_node_by_position(line, col)
         if node is None:
-            stack = inspect.stack()
```
Comment (goanpeca) [inline spyder/plugins/editor/extensions/snippets.py]: Why is this code removed @impact27

File: spyder/plugins/editor/panels/scrollflag.py [hunk 3]
```diff
@@ -82,6 +90,37 @@ def update_flag_colors(self, color_dict):
...
+            return
+
+        self._update_list_timer = QTimer(self)
+        self._update_list_timer.setSingleShot(True)
+        self._update_list_timer.timeout.connect(self.update_flags)
+        self._update_list_timer.start(REFRESH_RATE)
+
+    def update_flags(self):
+        """Update flags list."""
+        self._update_list_timer = None
```
Comment (goanpeca) [inline spyder/plugins/editor/panels/scrollflag.py]: Why to we make the timer none and then construct it. Could we not just create it once and use single shots and call start() every delayed update flags call?

Or even better not have that call at all?

File: spyder/plugins/editor/utils/debugger.py [hunk 4]
```diff
@@ -124,6 +124,8 @@ def clear_breakpoints(self):
         for data in self.editor.blockuserdata_list():
             data.breakpoint = False
             # data.breakpoint_condition = None  # not necessary, but logical
+        self.editor.sig_breakpoints_changed.emit()
```
Comment (goanpeca) [inline spyder/plugins/editor/utils/debugger.py]: Please add a comment @impact27 of why is this now needed, or why it was missing before

File: spyder/plugins/editor/utils/decoration.py [hunk 5]
```diff
@@ -58,7 +58,7 @@ def add(self, decorations):
             self.update()
         return added
 
-    def remove(self, decoration):
+    def remove(self, decoration, update=True):
         """
```
Comment (goanpeca) [inline spyder/plugins/editor/utils/decoration.py]: @impact27 please add an explanation (as long and complete as necessary) of what update does, and how it is used.

File: spyder/plugins/editor/panels/scrollflag.py [hunk 6]
```diff
@@ -82,6 +90,37 @@ def update_flag_colors(self, color_dict):
...
+        while block.isValid():
+            data = block.userData()
+            if data:
+                if data.code_analysis:
+                    self._code_analysis_list.append((block, data))
+                if data.todo:
+                    self._todo_list.append((block, data))
+                if data.breakpoint:
+                    self._breakpoint_list.append((block, data))
+            block = block.next()
```
Comment (goanpeca) [inline spyder/plugins/editor/panels/scrollflag.py]: for readability (since the if are not if-elif....) I suggest to change to

```python

                if data.code_analysis:
                    self._code_analysis_list.append((block, data))

                if data.todo:
                    self._todo_list.append((block, data))

                if data.breakpoint:
                    self._breakpoint_list.append((block, data))
```

File: spyder/utils/syntaxhighlighters.py [hunk 7]
```diff
@@ -539,6 +541,7 @@ def highlight_match(self, text, match, key, value, offset,
                     oedata.def_type = OutlineExplorerData.CELL
                     def_name = get_code_cell_name(text)
                     oedata.def_name = def_name
+                    self.sig_new_cell.emit(oedata)
```
Comment (goanpeca) [inline spyder/utils/syntaxhighlighters.py]: Please add a comment why this is now needed.

File: spyder/plugins/outlineexplorer/api.py [hunk 8]
```diff
@@ -48,6 +49,23 @@ def document_cells(block, forward=True):
     else:
         block = block.previous()
 
+    if not block.isValid():
+        return
+
+    if cell_list is not None:
```
Comment (goanpeca) [inline spyder/plugins/outlineexplorer/api.py]: Please add a comment or update the docstring for what cell_list does. I guess it is the caching you mentioned?

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 9]
```diff
@@ -612,6 +614,9 @@ def __init__(self, parent=None):
         completion_parent = self.completion_widget.parent()
         self.kite_call_to_action = KiteCallToAction(self, completion_parent)
 
+        # Some events should not be triggered on undo/redo
```
Comment (goanpeca) [inline spyder/plugins/editor/widgets/codeeditor.py]: Which type of events? Could you extend a bit here or add examples so we now what those events are or could be in the future?

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 10]
```diff
@@ -1574,7 +1579,7 @@ def set_language(self, language, filename=None):
                     self.comment_string = comment_string
                     if key in CELL_LANGUAGES:
                         self.supported_cell_language = True
-                        self.cell_separators = CELL_LANGUAGES[key]
```
Comment (goanpeca) [inline spyder/plugins/editor/widgets/codeeditor.py]: I do not understand why this change was needed @impact27

File: spyder/plugins/editor/extensions/snippets.py (hunk 2 above)
Comment (goanpeca) [reply spyder/plugins/editor/extensions/snippets.py]: @andfoy some context would help :-)

File: spyder/plugins/editor/widgets/base.py [hunk 11]
```diff
@@ -380,6 +380,24 @@ def keyPressEvent(self, event):
             super(TextEditBaseWidget, self).keyPressEvent(event)
 
     #------Text: get, set, ...
+    def add_to_cell_list(self, oedata):
+        """Add new cell to cell list."""
+        self._cell_list.append(oedata)
```
Comment (goanpeca) [inline spyder/plugins/editor/widgets/base.py]: What happens when cells are removed? is this also being updated somewhere or we just keep accumulating information on this variable?

File: spyder/plugins/editor/widgets/codeeditor.py (hunk 10 above)
Comment (goanpeca) [reply spyder/plugins/editor/widgets/codeeditor.py]: We do not need to know the separators but just that is supports cells?

File: spyder/plugins/editor/panels/scrollflag.py [hunk 12]
```diff
@@ -108,58 +147,104 @@ def paintEvent(self, event):
...
-                    painter.drawRect(rect_x, rect_y, rect_w, rect_h)
-                if data.breakpoint:
-                    # Paint the breakpoints
-                    rect_y = self.calcul_flag_ypos(
-                        line_number, scale_factor, offset)
-                    painter.setBrush(self._facecolors['breakpoint'])
-                    painter.setPen(self._edgecolors['breakpoint'])
-                    painter.drawRect(rect_x, rect_y, rect_w, rect_h)
-            block = block.next()
+        editor = self.editor
```
Comment (goanpeca) [inline spyder/plugins/editor/panels/scrollflag.py]: @impact27 I am having a really hard time following all these changes, please add comments on each block to explain what it does

File: spyder/plugins/editor/panels/scrollflag.py [hunk 13]
```diff
@@ -108,58 +147,104 @@ def paintEvent(self, event):
...
+        print_local = not bool(self.slider)
+
+        if not print_local:
```
Comment (goanpeca) [inline spyder/plugins/editor/panels/scrollflag.py]: For instance, it seems we do not need to compute stuff if the slider scrollbar are is off?

But the code itslef is not enough we really need to document this with comments.

File: spyder/plugins/editor/panels/scrollflag.py [hunk 14]
```diff
@@ -108,58 +147,104 @@ def paintEvent(self, event):
...
+                last_line + 0.5, scale_factor, offset) - self.FLAGS_DY / 2
+
+            def calcul_flag_ypos(block):
+                line_number = block.firstLineNumber()
+                frac = line_number / last_line
+                pos = first_y_pos + frac * (last_y_pos - first_y_pos)
+                return ceil(pos)
+
+        else:
+            # Only print visible flags
```
Comment (goanpeca) [inline spyder/plugins/editor/panels/scrollflag.py]: Print? Was it more a Paint?

File: spyder/plugins/editor/widgets/base.py [hunk 15]
```diff
@@ -454,82 +472,22 @@ def get_cell_as_executable_code(self, cursor=None):
             text = ls * line_from + text
         return text, block
 
-    def is_cell_separator(self, cursor=None, block=None):
-        """Return True if cursor (or text block) is on a block separator"""
```
Comment (goanpeca) [inline spyder/plugins/editor/widgets/base.py]: Why was all this removed? 

select_current_cell is now doing what select_current_cell_in_visible_portion was or ?

File: spyder/plugins/editor/widgets/base.py [hunk 16]
```diff
@@ -454,82 +472,22 @@ def get_cell_as_executable_code(self, cursor=None):
...
         returns
          -the textCursor
          -a boolean indicating if the entire file is selected
-         """
-
-        cursor = self.textCursor()
+        """
+        if cursor is None:
+            cursor = self.textCursor()
         if self.current_cell:
```
Comment (goanpeca) [inline spyder/plugins/editor/widgets/base.py]: Add space between if's for readability

File: spyder/plugins/editor/widgets/base.py [hunk 17]
```diff
@@ -563,39 +525,32 @@ def select_current_cell_in_visible_portion(self):
...
-        while not self.is_cell_separator(cursor):
-            # Moving to the next code cell
-            cursor.movePosition(QTextCursor.NextBlock)
-            prev_pos = cur_pos
-            cur_pos = cursor.position()
-            if cur_pos == prev_pos:
-                return
+        block = cursor.block()
+        try:
+            footer = next(document_cells(
```
Comment (goanpeca) [inline spyder/plugins/editor/widgets/base.py]: Is this to avoid the while loop or?

File: spyder/plugins/editor/extensions/snippets.py (hunk 2 above)
Comment (impact27) [reply spyder/plugins/editor/extensions/snippets.py]: The code here want to know if it is called from "undo" or "redo". It does that by inspecting the call stack, which is not a very nice way to do this as it requires loading the source files for the entire call stack! I therefore changed that to add a `self.editor.is_undoing` to check for undo and redos instead

File: spyder/plugins/editor/panels/scrollflag.py (hunk 3 above)
Comment (impact27) [reply spyder/plugins/editor/panels/scrollflag.py]: > Why to we make the timer none and then construct it. Could we not just create it once and use single shots and call start() every delayed update flags call?

Can I reuse a timer?

> Or even better not have that call at all?

I am not sure I understand what you suggest. The flags needs to be updated at some point surely?

File: spyder/plugins/editor/widgets/base.py (hunk 11 above)
Comment (impact27) [reply spyder/plugins/editor/widgets/base.py]: Line 394 filters the list every call

File: spyder/plugins/editor/widgets/base.py (hunk 15 above)
Comment (impact27) [reply spyder/plugins/editor/widgets/base.py]: Yes `select_current_cell_in_visible_portion` was not actually doing something different from `select_current_cell` and `is_cell_separator` was duplicated code as well.

File: spyder/plugins/editor/widgets/base.py (hunk 17 above)
Comment (impact27) [reply spyder/plugins/editor/widgets/base.py]: Yes we just access the next cell instead of reading every line until we reach a cell.

File: spyder/plugins/editor/widgets/codeeditor.py (hunk 9 above)
Comment (impact27) [reply spyder/plugins/editor/widgets/codeeditor.py]: see discussion about `cursor_changed`.

File: spyder/plugins/editor/widgets/codeeditor.py (hunk 10 above)
Comment (impact27) [reply spyder/plugins/editor/widgets/codeeditor.py]: Because the whole cell logic was separated between two systems: the syntax highlighter that keeps a list of cell for efficient cell finding, and the editor that reads multiple times through the whole file to find cells.

Comment (impact27) [reply spyder/plugins/editor/widgets/codeeditor.py]: So I removed the second system as it was causing performances issues on large files.

File: spyder/plugins/editor/panels/scrollflag.py (hunk 6 above)
Comment (impact27) [reply spyder/plugins/editor/panels/scrollflag.py]: Done

File: spyder/plugins/editor/panels/scrollflag.py (hunk 12 above)
Comment (impact27) [reply spyder/plugins/editor/panels/scrollflag.py]: Done

File: spyder/plugins/editor/panels/scrollflag.py (hunk 13 above)
Comment (impact27) [reply spyder/plugins/editor/panels/scrollflag.py]: Done

File: spyder/plugins/editor/utils/debugger.py (hunk 4 above)
Comment (impact27) [reply spyder/plugins/editor/utils/debugger.py]: Done. I don't know why it was missing but as the flags were recomputed completely for every print event (!!), this didn't matter.

File: spyder/plugins/editor/utils/decoration.py (hunk 5 above)
Comment (impact27) [reply spyder/plugins/editor/utils/decoration.py]: Done

File: spyder/plugins/editor/widgets/base.py (hunk 16 above)
Comment (impact27) [reply spyder/plugins/editor/widgets/base.py]: Done

File: spyder/plugins/outlineexplorer/api.py (hunk 8 above)
Comment (impact27) [reply spyder/plugins/outlineexplorer/api.py]: Done. Yes it is

File: spyder/utils/syntaxhighlighters.py (hunk 7 above)
Comment (impact27) [reply spyder/utils/syntaxhighlighters.py]: Done

File: spyder/plugins/editor/panels/scrollflag.py (hunk 3 above)
Comment (goanpeca) [reply spyder/plugins/editor/panels/scrollflag.py]: >Can I reuse a timer?

Sure, you start it as many times as you want.

File: spyder/plugins/editor/utils/debugger.py (hunk 4 above)
Comment (goanpeca) [reply spyder/plugins/editor/utils/debugger.py]: Understood!

File: spyder/plugins/editor/widgets/base.py (hunk 11 above)
Comment (goanpeca) [reply spyder/plugins/editor/widgets/base.py]: 👍

File: spyder/plugins/editor/widgets/base.py (hunk 15 above)
Comment (goanpeca) [reply spyder/plugins/editor/widgets/base.py]: Got it!

File: spyder/plugins/editor/widgets/base.py (hunk 17 above)
Comment (goanpeca) [reply spyder/plugins/editor/widgets/base.py]: Cool :+1:

File: spyder/plugins/editor/widgets/codeeditor.py (hunk 10 above)
Comment (goanpeca) [reply spyder/plugins/editor/widgets/codeeditor.py]: Understood

File: spyder/plugins/outlineexplorer/api.py (hunk 8 above)
Comment (goanpeca) [reply spyder/plugins/outlineexplorer/api.py]: Thanks!

File: spyder/plugins/editor/extensions/snippets.py (hunk 2 above)
Comment (goanpeca) [reply spyder/plugins/editor/extensions/snippets.py]: Understood!

File: spyder/plugins/editor/widgets/codeeditor.py (hunk 9 above)
Comment (goanpeca) [reply spyder/plugins/editor/widgets/codeeditor.py]: 👍

File: spyder/plugins/editor/extensions/snippets.py [hunk 18]
```diff
@@ -700,17 +700,9 @@ def cursor_changed(self, line, col):
...
-            ignore = False
-            # Check if parent call was due to text update on codeeditor
-            # caused by a undo/redo call
-            for call in stack:
-                if call[3] in ignore_calls:
-                    ignore = True
-                    break
+            ignore = self.editor.is_undoing or self.editor.is_redoing
```
Comment (andfoy) [inline spyder/plugins/editor/extensions/snippets.py]: How are race conditions managed with these variables? i.e., There could be the case where the editor flags are set to False and this slot might not be called at all

Comment (impact27) [reply spyder/plugins/editor/extensions/snippets.py]: If this is called from the main thread there is no race condition. Can this be called from another thread?

**Comment by ccordoba12** [comment]
@goanpeca, please give your final approval to this one if you it's ready.

File: spyder/plugins/editor/extensions/snippets.py (hunk 18 above)
Comment (andfoy) [reply spyder/plugins/editor/extensions/snippets.py]: Well, personally, I don't know much about Signal dispatching in Qt, so if you say that everything occurs sequentially, I trust you

File: spyder/plugins/editor/panels/scrollflag.py [hunk 19]
```diff
@@ -108,58 +164,111 @@ def paintEvent(self, event):
...
-                        line_number, scale_factor, offset)
-                    painter.setBrush(self._facecolors['breakpoint'])
-                    painter.setPen(self._edgecolors['breakpoint'])
-                    painter.drawRect(rect_x, rect_y, rect_w, rect_h)
-            block = block.next()
+        editor = self.editor
+        # Check if the slider is visible
+        paint_local = not bool(self.slider)
+
+        # Define calcul_flag_ypos to position the flags:
```
Comment (ccordoba12) [inline spyder/plugins/editor/panels/scrollflag.py]: Please change `calcul_flag_ypos` to `compute_flag_ypos`

File: spyder/plugins/editor/panels/scrollflag.py [hunk 20]
```diff
@@ -108,58 +164,111 @@ def paintEvent(self, event):
...
+            # Paint flags for the entire document
+            last_line = editor.document().lastBlock().firstLineNumber()
+            # The 0.5 offset is used to align the flags with the center of
+            # their corresponding text edit block before scaling.
+            first_y_pos = self.value_to_position(
+                0.5, scale_factor, offset) - self.FLAGS_DY / 2
+            last_y_pos = self.value_to_position(
+                last_line + 0.5, scale_factor, offset) - self.FLAGS_DY / 2
+
+            def calcul_flag_ypos(block):
```
Comment (ccordoba12) [inline spyder/plugins/editor/panels/scrollflag.py]: `calcul_flag_ypos` -> `compute_flag_ypos`

File: spyder/plugins/editor/panels/scrollflag.py [hunk 21]
```diff
@@ -108,58 +164,111 @@ def paintEvent(self, event):
...
+        else:
+            # Only paint flags for visible lines
+            visible_lines = [val[1] for val in editor.visible_blocks]
+            if not visible_lines:
+                # Nothing to do
+                return
+            min_line = min(visible_lines)
+            max_line = max(visible_lines)
+
+            def calcul_flag_ypos(block):
```
Comment (ccordoba12) [inline spyder/plugins/editor/panels/scrollflag.py]: `calcul_flag_ypos` to `compute_flag_ypos`

File: spyder/plugins/editor/extensions/snippets.py (hunk 18 above)
Comment (impact27) [reply spyder/plugins/editor/extensions/snippets.py]: Usually, when a signal is emitted from the main thread, it is processed immediately. This is why the call inspection worked as well because this needs to be called directly to have “undo” in the stack.
